### PR TITLE
UPSTREAM: <drop>: fixup for 14537

### PR DESCRIPTION
**Pull request by markturansky**
Conflicts and rejected patch files means I missed a few things in the origin patch PR.  

This PR cleans up a few issues and fixes one bug.

@deads2k @pweil-

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

**Comment by markturansky** [comment]
Unrelated flake?

```
Error from server: 501: All the given peers are not reachable (failed to propose on members [https://172.18.15.99:4001] twice [last error: Unexpected HTTP status code]) [0]
!!! Error in hack/../test/end-to-end/core.sh:164
	'oc delete pod cli-with-token-2' exited with status 1
Call stack:
	1: hack/../test/end-to-end/core.sh:164 main(...)
Exiting with status 1
Error from server: User "e2e-user" cannot list all buildconfigs in the cluster
```

**Comment by pweil-** [comment]
re[test]

**Comment by markturansky** [comment]
Different flake:

```
Error from server: User "system:anonymous" cannot list pods in project "default"
Dry run enabled - no modifications will be made. Add --confirm to remove images
!!! Error in hack/../test/cmd/../../hack/cmd_util.sh:407
	'return 1' exited with status 1
Call stack:
	1: hack/../test/cmd/../../hack/cmd_util.sh:407 os::cmd::try_until_success(...)
	2: hack/../test/cmd/newapp.sh:20 main(...)
Exiting with status 1
!!! Error in hack/test-cmd.sh:284
	'${test}' exited with status 1
Call stack:
	1: hack/test-cmd.sh:284 main(...)
Exiting with status 1
make: *** [test-int-plus] Error 1
```

**Comment by markturansky** [comment]
[test]

**Comment by ncdc** [comment]
@markturansky FYI that wasn't the flake. The flake was https://github.com/openshift/origin/issues/6259#issuecomment-165704507, which we're trying to fix with #6407

**Comment by markturansky** [comment]
Thanks for the clarification.

**Comment by ncdc** [comment]
Latest failure:

```
Running hack/../test/cmd/builds.sh:79: executing 'oc patch bc/ruby-sample-build -p '{"spec":{"output":{"to":{"name":"different:tag1"}}}}'' expecting success...
FAILURE after 0.279s: hack/../test/cmd/builds.sh:79: executing 'oc patch bc/ruby-sample-build -p '{"spec":{"output":{"to":{"name":"different:tag1"}}}}'' expecting success: the command returned the wrong error code
There was no output from the command.
Standard error from the command:
Error from server: unable to find api field in struct BuildConfig for the json field "metadata"
!!! Error in hack/../test/cmd/../../hack/cmd_util.sh:193
	'return 1' exited with status 1
Call stack:
	1: hack/../test/cmd/../../hack/cmd_util.sh:193 os::cmd::expect_success(...)
	2: hack/../test/cmd/builds.sh:79 main(...)
Exiting with status 1
!!! Error in hack/test-cmd.sh:284
	'${test}' exited with status 1
Call stack:
	1: hack/test-cmd.sh:284 main(...)
Exiting with status 1
[FAIL] !!!!! Test Failed !!!!
```

[test]

**Comment by ncdc** [comment]
:snowflake: :snowman: 

```
Error from server: 501: All the given peers are not reachable (failed to propose on members [https://172.18.3.40:4001] twice [last error: Unexpected HTTP status code]) [0]
!!! Error in hack/../test/end-to-end/core.sh:164
	'oc delete pod cli-with-token-2' exited with status 1
Call stack:
	1: hack/../test/end-to-end/core.sh:164 main(...)
Exiting with status 1

[FAIL] !!!!! Test Failed !!!!
```

[test]

**Comment by markturansky** [comment]
woot!  all green!

**Comment by danmcp** [comment]
@ncdc I think you asked me about it earlier.  It isn't tagged for merge though.

**Comment by smarterclayton** [comment]
Going to merge so it makes a build afterwards but I want to see a LGTM and sign off from the upstream reviewer. [merge]

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go [hunk 1]
```diff
@@ -139,13 +156,42 @@ func (binder *PersistentVolumeClaimBinder) addClaim(obj interface{}) {
 func (binder *PersistentVolumeClaimBinder) updateClaim(oldObj, newObj interface{}) {
 	binder.lock.Lock()
 	defer binder.lock.Unlock()
-	newClaim := newObj.(*api.PersistentVolumeClaim)
-	err := syncClaim(binder.volumeIndex, binder.client, newClaim)
-	if err != nil {
+	newClaim, ok := newObj.(*api.PersistentVolumeClaim)
+	if !ok {
+		glog.Errorf("Expected PersistentVolumeClaim but handler received %+v", newObj)
```
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go]: shouldn't this be a return? you go on to use `newClaim` anyway, which seems wrong

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go [hunk 2]
```diff
@@ -139,13 +156,42 @@ func (binder *PersistentVolumeClaimBinder) addClaim(obj interface{}) {
...
+				volume = pv
+			}
+		}
+	}
+	if unk, ok := obj.(cache.DeletedFinalStateUnknown); ok && unk.Obj != nil {
+		if pv, ok := unk.Obj.(*api.PersistentVolume); ok {
+			volume = pv
+		}
+	}
+
```
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go]: what happens if you get here and volume was never set?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 3]
```diff
@@ -73,15 +73,20 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
...
 		},
 		&api.PersistentVolume{},
-		// TODO: Can we have much longer period here?
 		syncPeriod,
 		framework.ResourceEventHandlerFuncs{
 			AddFunc: func(obj interface{}) {
-				pv := obj.(*api.PersistentVolume)
+				pv, ok := obj.(*api.PersistentVolume)
+				if !ok {
+					glog.Errorf("Error casting object to PersistentVolume: %v", obj)
```
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: return, don't use `pv`?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/persistentvolume_recycler_controller.go [hunk 4]
```diff
@@ -73,15 +73,20 @@ func NewPersistentVolumeRecycler(kubeClient client.Interface, syncPeriod time.Du
...
+				}
 				recycler.reclaimVolume(pv)
 			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
-				pv := newObj.(*api.PersistentVolume)
+				pv, ok := newObj.(*api.PersistentVolume)
+				if !ok {
+					glog.Errorf("Error casting object to PersistentVolume: %v", newObj)
```
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/persistentvolume_recycler_controller.go]: return

**Pull request merged**